conda/core/solve/state.py

```diff
@@ -0,0 +1,717 @@
+            if name not in pin_overrides:
+                self.specs.set(name, spec, reason="Explicitly requested by user")
+
+        ### Conda pinning ###
+
+        # As a business rule, we never want to downgrade conda below the current version,
+        # unless that's requested explicitly by the user (which we actively discourage).
+
+        if "conda" in self.specs and paths_equal(sis.prefix, context.conda_prefix) and "conda" in sis.installed:
+            record = sis.installed["conda"]
+            spec = self.specs["conda"]
+            required_version = f">={record.version}"
+            if not spec.get("version"):
+                spec = MatchSpec(spec, version=required_version)
+                reason = "Pinning conda with version greater than currently installed"
+            if context.auto_update_conda and "conda" not in sis.requested:
+                spec = MatchSpec("conda", version=required_version, target=None)
+                reason = "Pinning conda with version greater than currently installed, auto update"
+            self.specs.set("conda", spec, reason=reason)
+
+        ### Extra logic ###
+        # this is where we are adding workarounds for mamba difference's in behavior, which might not belong
+        # here as they are solver specific
+
+        # next step -> .prepare_for_solve()
+
+
+    def prepare_for_remove(self) -> dict[str, MatchSpec]:
+        pass
+
+
+    def prepare_for_solve(self):
+        ### Inconsistency analysis ###
+        # here we would call conda.core.solve.classic.Solver._find_inconsistent_packages()
+
+        ### Conflict minimization ###
+        # here conda.core.solve.classic.Solver._run_sat() enters a `while conflicting_specs`
+        # loop to neuter some of the specs in self.specs. In other solvers we let the solver run into them.
+        # We might need to add a hook here ?
+
+        # After this, we finally let the solver do its work. It will either finish with a final state
+        # or fail and repopulate the conflicts list in the SolverOutputState object
+        ...
+
+    def post_solve(self):
+        # After a solve, we still need to do some refinement
+
+        ### Neutered ###
+        # annotate overridden history specs so they are written to disk
+
+        ### Add inconsistent packages back ###
+        # direct result of the inconsistency analysis above
+
+        ### Deps modifier ###
+        # handle the different modifiers (NO_DEPS, ONLY_DEPS, UPDATE_DEPS)
+        # this might mean removing different records by hand or even calling the solver a 2nd time
+
+        ### Prune ###
+        # remove orphan leaves in the graph
+        ...
+
+    @property
+    def _pools(self):
+        # ordered by priority (high to low)
+        # note this is only an approximation
+        return (
+            self._explicit,
+            self._hard_constrained,
+            self._soft_constrained,
+            self._unconstrained,
+        )
+
+    @functools.lru_cache
```

Comment:
I think we can use `functools.cache` here, or for Python<3.9 compat:
```suggestion
    @functools.lru_cache(maxsize=None)
```